src/riscv/lib/src/jit.rs

```diff
@@ -115,40 +115,54 @@ impl JIT {
         })
     }
 
-    /// Compile a sequence of instructions to a callable native function.
+    /// Compile instructions in a page starting from an entrypoing at the given offest.
+    /// Continue compiling instructions of either `Next` or `Branch` outcomes.
     ///
     /// Not all instructions are currently supported. For blocks containing
     /// unsupported instructions, `None` will be returned.
-    pub fn compile<D, MC: MemoryConfig>(
+    pub fn compile_page<D, MC: MemoryConfig>(
         &mut self,
-        instr: &[Instruction],
+        page: &JittedPage<D, MC>,
+        mut offset: usize,
         program_counter: Address,
     ) -> Option<JitFn<D, MC>> {
-        let Ok(hash) = Hash::blake3_hash((instr, program_counter)) else {
-            return None;
-        };
-
         let mut builder = self.start::<D, MC>(program_counter);
-        let mut lowered_instrs = Vec::with_capacity(instr.len());
+        let mut lowered_instrs = Vec::new();
+        let mut instr_for_hash = Vec::new();
+
+        // Compile instructions starting at the given offset until we encounter an instruction that
+        // does not continue with either a `Next` or `Branch` outcome, or we reach the end of the page.
+        let mut should_continue = true;
+        while should_continue && offset < page.entries.len() {
+            let instr = page.entries[offset].as_ref();
 
-        // Check if the opcode of the instruction is supported in JIT and stop compilation in JIT if not.
-        for i in instr {
-            let Some(lower) = i.opcode.to_lowering() else {
+            let Some(lower) = instr.opcode.to_lowering() else {
                 builder.abandon();
                 self.clear();
                 return None;
             };
 
-            let mut instr_builder = builder.build_next_instruction(i.width());
+            let mut instr_builder = builder.build_next_instruction(instr.width());
 
             let instr_result = unsafe {
                 // # SAFETY: lower is called with args from the same instruction that it
                 // was derived
-                (lower)(i.args(), &mut instr_builder)
+                (lower)(instr.args(), &mut instr_builder)
             };
 
             let lowered_instr = instr_builder.finish(instr_result);
+            should_continue = matches!(
+                lowered_instr.outcomes(),
+                InstructionOutcomes::Next { .. } | InstructionOutcomes::Branch { .. }
+            );
+
             lowered_instrs.push(lowered_instr);
+
+            instr_for_hash.push(*instr);
+
+            if should_continue {
+                offset += (instr.width() as usize) >> 1;
+            }
```

Comment:
Nit: The additional state required across loop iterations can be avoided by using the Rust native `break` keyword. This makes it very clear where the iteration stops.

```suggestion
            let should_continue = matches!(
                lowered_instr.outcomes(),
                InstructionOutcomes::Next { .. } | InstructionOutcomes::Branch { .. }
            );

lowered_instrs.push(lowered_instr);

            instr_for_hash.push(*instr);

            if !should_continue {
                break;
            }

            offset += (instr.width() as usize) >> 1;
```